README.md
# keyboard_shortcuts

keyboard_shortcuts allows some functionality to be handled with the keyboard.
It adds an icon to the mail screen to view all available shortcuts. 

## History

Cor Bosman created the project and moved it to https://github.com/corbosman/keyboard_shortcuts

With this fork, the latest updates of roundcube are applied to style, and some small extension of shortcuts.
It is tested with roundcube version 1.4.9 (October 2020).

## Contact

Bug reports through github (https://github.com/corbosman/keyboard_shortcuts/issues)
or https://github.com/mattanja/keyboard_shortcuts/issues

## LICENSE

This plugin is distributed under the GNU General Public License Version 2.
Please read through the file LICENSE for more information about this license.


CHANGELOG
CHANGELOG keyboard_shortcuts
============================

RELEASE 2.4.2

- fix bugs caused by obsolete core $rcmail class methods
- include version info in composer.json
- copy image file to elastic skin

RELEASE 2.4.1

- fix composer.json

RELEASE 2.4

- change how we detect focus
- prevent other keyboard events from firing

RELEASE 2.3

- added extra check on input fields (fix by Rosali)
- fixed Select All Visible keyboard shortcut (fix by Rodrigo Castillo)

RELEASE 2.1

- added Russian translation
- bugfix: j/k keys were documented wrong

RELEASE 2.0
-----------

- move to github
- bugfix: dont fire when user is holding a meta key
- bugfix: many browser can bind to both body and document, dont run both. $(document) should work everywhere
- bug:fix we can be in any input element, make sure we dont interfere



config.inc.php
<?php

/**
 *  register shortcut commands
 *
 */

/**
 * user configurable?
 */
$rcmail_config['keyboard_shortcuts_userconfigurable'] = true;

/**
 * checking if the server supports threads is expensive. Just tell me.
 */
$rcmail_config['keyboard_shortcuts_threads'] = true;

/**
 * default key bindings
 */
$rcmail_config['keyboard_shortcuts_default'] = array(
    'all' => array(
        'F1' => 'mail',
        'F2' => 'addressbook',
        'F3' => 'settings',
        'F8' => 'help',
        's' => 'search',
    ),

    'mail' => array(
        'a' => 'select-page',
        'c' => 'compose',
        'd' => 'delete',
        'f' => 'forward',
        'i' => 'select-invert',
        'j' => 'previouspage',
        'k' => 'nextpage',
        'p' => 'print',
        'r' => 'reply',
        'Shift+r' => 'replyall',
        'Ctrl+r' => 'mark-read',
        'Ctrl+u' => 'mark-unread',
        'x' => 'select-none',
    ),
    'addressbook' => array(
    ),
    'compose' => array(
    )
);

/**
 * if we support threads, add some extra default keys
 */
if($rcmail_config['keyboard_shortcuts_threads'] === true) {
    $rcmail_config['keyboard_shortcuts_default']['mail']['Shift+c'] = 'collapse-all';
    $rcmail_config['keyboard_shortcuts_default']['mail']['Shift+e'] = 'expand-all';
    $rcmail_config['keyboard_shortcuts_default']['mail']['Shift+u'] = 'expand-unread';
}

commands.inc.php
<?php

/**
 * All the commands and their options. Each command is listed in the screen section they're active in.
 *
 * options:
 *
 * function     the function to call, default = name of command
 * parameters   any parameters to the function we call, can be string or array
 * label        a label for this command, default = name of command
 */

$rcmail_config['keyboard_shortcuts_commands'] = array(
    'all'           => array(
            'addressbook'   => array(),
            'help'          => array(),
            'logout'        => array(),
            'mail'          => array(),
            'settings'      => array(),
            'search'        => array()
    ),
    'mail'          => array(
            'compose'           => array(),
            'delete'            => array(),
            'download'          => array(),
            'forward'           => array(),
            'mark-flagged'      => array('function' => 'mark', 'parameters' => 'flagged'),
            'mark-read'         => array('function' => 'mark', 'parameters' => 'read'),
            'mark-unflagged'    => array('function' => 'mark', 'parameters' => 'unflagged'),
            'mark-unread'       => array('function' => 'mark', 'parameters' => 'unread'),
            'nextpage'          => array(),
            'previouspage'      => array(),
            'print'             => array(),
            'reply'             => array(),
            'replyall'          => array(),
            'select-all'        => array(),
            'select-flagged'    => array('function' => 'select-all', 'parameters' => 'flagged'),
            'select-invert'     => array('function' => 'select-all', 'parameters' => 'invert'),
            'select-none'       => array('function' => 'select-all', 'parameters' => 'none'),
            'select-page'       => array('function' => 'select-all', 'parameters' => 'page'),
            'select-unread'     => array('function' => 'select-all', 'parameters' => 'unread'),
            'viewsource'        => array()
    ),
    'addressbook'   => array(),
    'compose'       => array(),
);

/**
 * if we support threads, add some extra commands
 */
if(isset($rcmail_config['keyboard_shortcuts_threads']) and $rcmail_config['keyboard_shortcuts_threads'] === true) {
    $rcmail_config['keyboard_shortcuts_commands']['mail']['collapse-all'] = array();
    $rcmail_config['keyboard_shortcuts_commands']['mail']['expand-all'] = array();
    $rcmail_config['keyboard_shortcuts_commands']['mail']['expand-unread'] = array();
}


keyboard_shortcuts.js
function keyboard_shortcuts_show_help() {
  $('#keyboard_shortcuts_help').dialog('open');
}

$(function() {

  // initialize a dialog window
  $('#keyboard_shortcuts_help').dialog({
    autoOpen: false,
    draggable: true,
    modal: false,
    resizable: false,
    width: 750,
    title: rcmail.gettext("keyboard_shortcuts.keyboard_shortcuts")
  });

  // fire up the keypress event listener
  $(document).keypress(function (e) {
    return key_pressed(e);
  });


  function key_pressed (e) {
    // special case. If we hit ctrl-enter, and we're composing, and we have focus, then send email
    if (rcmail.env.action == 'compose' && e.which == 13 && e.ctrlKey && $("*:focus").is("#composebody")) {
      $('.button.send').click();
      return false;
    }

    // check if some element has focus. If it does, skip this plugin.
    if ( $("*:focus").is("textarea, input") ) return true;

    if (rcmail.env.action == 'compose' || rcmail.env.task == 'login' || e.ctrlKey || e.metaKey) return true;

    if (rcmail.env.action == '') {	// list mailbox

      if(rcmail.env.ks_functions[e.which]) {
        this[rcmail.env.ks_functions[e.which]]();
        return false;
      }

      switch (e.which) {
          case 63:		// ? = help
          //keyboard_shortcuts_show_help();
          var ks_function = rcmail.env.ks_functions[e.which];
          this[ks_function]();

          return false;
        case 65:		// A = mark all as read
          rcmail.command('select-all', 'page');
          rcmail.command('mark', 'read');
          return false;
        case 67:                // C = collapse-all
          rcmail.command('collapse-all');
          return false;
        case 69:                // E = expand-all
          rcmail.command('expand-all');
          return false;
        case 82:		// R = reply-all
          if (rcmail.message_list.selection.length == 1)
          rcmail.command('reply-all');
          return false;
        case 85:                // U = expand-unread
          rcmail.command('expand-unread');
          return false;
        case 97:		// a = select all
          rcmail.command('select-all', 'page');
          return false;
        case 99:		// c = compose
          rcmail.command('compose');
          return false;
        case 100:		// d = delete
          rcmail.command('delete', '', rcmail);
          return false;
        case 102:		// f = forward
          if (rcmail.message_list.selection.length == 1)
          rcmail.command('forward');
          return false;
        case 106:		// j = previous page (similar to Gmail)
          rcmail.command('previouspage');
          return false;
        case 107:		// k = next page (similar to Gmail)
          rcmail.command('nextpage');
          return false;
        case 112:		// p = print
          if (rcmail.message_list.selection.length == 1)
          rcmail.command('print');
          return false;
        case 114:		// r = reply
          if (rcmail.message_list.selection.length == 1)
          rcmail.command('reply');
          return false;
        case 115:		// s = search
          $('#quicksearchbox').focus();
          $('#quicksearchbox').select();
          return false;
        case 117:		// u = update (check for mail)
          rcmail.command('checkmail');
          return false;
        case 101:   // e = archive (similar to Gmail)
        case 122:		// z = archive (move to archive)
          rcmail.command('plugin.archive');
          return false;
      }
    } else if (rcmail.env.action == 'show' || rcmail.env.action == 'preview') {
      switch (e.which) {
        case 82:		// R = reply-all
          rcmail.command('reply-all');
          return false;
        case 99:		// c = compose
          rcmail.command('compose');
          return false;
        case 100:		// d = delete
          rcmail.command('delete');
          return false;
        case 102:		// f = forward
          rcmail.command('forward');
          return false;
        case 105:		// i = back to list
          rcmail.command('list');
          return false;
        case 106:		// j = previous message (similar to Gmail)
          rcmail.command('previousmessage');
          return false;
        case 107:		// k = next message (similar to Gmail)
          rcmail.command('nextmessage');
          return false;
        case 112:		// p = print
          rcmail.command('print');
          return false;
        case 114:		// r = reply
          rcmail.command('reply');
          return false;
        case 122:		// z = archive (move to archive)
          rcmail.command('plugin.archive');
          return false;
      }
    }
  }
});

// support functions for each function we support
function ks_help() {
  keyboard_shortcuts_show_help();
}


keyboard_shortcuts.php
<?php
/**
 * keyboard_shortcuts
 *
 * Enables some common tasks to be executed with keyboard shortcuts
 *
 * @version 2.4.2 2023-03-09
 * @author Patrik Kullman / Roland 'rosali' Liebl / Cor Bosman <roundcube@wa.ter.net>
 * @licence GNU General Public License, version 2
 * @link https://github.com/corbosman/keyboard_shortcuts
 *
 **/

/**
 * Shortcuts, list view:
 * ?:	Show shortcut help
 * a:	Select all visible messages
 * A:	Mark all as read (as Google Reader)
 * c:	Compose new message
 * d:	Delete message
 * f:	Forward message
 * j:	Go to previous page of messages (as Gmail)
 * k:	Go to next page of messages (as Gmail)
 * p:	Print message
 * r:	Reply to message
 * R:	Reply to all of message
 * s:	Jump to quicksearch
 * u:	Check for new mail (update)
 * z:	Move message to archive
 *
 * Shortcuts, threads view:
 * E:   Expand all
 * C:   Collapse all
 * U:   Expand Unread
 *
 * Shortcuts, mail view:
 * d:	Delete message
 * f:	Forward message
 * i:	Go to back to message list (as Gmail)
 * j:	Go to previous message (as Gmail)
 * k:	Go to next message (as Gmail)
 * p:	Print message
 * r:	Reply to message
 * R:	Reply to all of message
 * z:	Move message to archive
 */

class keyboard_shortcuts extends rcube_plugin
{
    public $task = 'mail|compose';

    function init()
    {
      // only init in authenticated state and if newuserdialog is finished
      // do not init on compose (css incompatibility with compose_addressbook plugin
      $rcmail = rcmail::get_instance();
      $this->require_plugin('jqueryui');

      if($_SESSION['username'] && empty($_SESSION['plugin.newuserdialog'])){
        $this->include_stylesheet('keyboard_shortcuts.css');
        $this->include_script('keyboard_shortcuts.js');
        $this->add_hook('template_container', array($this, 'html_output'));
        $this->add_texts('localization', true);
      }
    }

    function html_output($p) {
      if ($p['name'] == "listcontrols") {
        $rcmail = rcmail::get_instance();
        $skin  = $rcmail->config->get('skin');

        if(!file_exists('plugins/keyboard_shortcuts/skins/' . $skin . '/images/keyboard.png')){
          $skin = "default";
        }

        $this->load_config();
        $keyboard_shortcuts = $rcmail->config->get('keyboard_shortcuts_extras', array());
        $archive_supported = $rcmail->config->get('archive_mbox');

        $c = "";
        $c .= '<span id="keyboard_shortcuts_title">' . $this->gettext("title") . ":&nbsp;</span><a id='keyboard_shortcuts_link' href='#' class='button' title='".$this->gettext("keyboard_shortcuts")." ".$this->gettext("show")."' onclick='return keyboard_shortcuts_show_help()'><img align='top' src='plugins/keyboard_shortcuts/skins/".$skin."/images/keyboard.png' alt='".$this->gettext("keyboard_shortcuts")." ".$this->gettext("show")."' /></a>\n";
        $c .= "<div id='keyboard_shortcuts_help'>";
        $c .= "<div><h4>".$this->gettext("mailboxview")."</h4>";
        $c .= "<div class='shortcut_key'>?</div> ".$this->gettext('help')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>a</div> ".$this->gettext('selectallvisiblemessages')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>A</div> ".$this->gettext('markallvisiblemessagesasread')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>c</div> ".$this->gettext('compose')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>d</div> ".$this->gettext('deletemessage')."<br class='clear' />";
        if (!empty($archive_supported)) {
          $c .= "<div class='shortcut_key'>z</div> ".$this->gettext('archive.buttontitle')."<br class='clear' />";
        }
        $c .= "<div class='shortcut_key'>f</div> ".$this->gettext('forwardmessage')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>j</div> ".$this->gettext('previouspage')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>k</div> ".$this->gettext('nextpage')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>p</div> ".$this->gettext('printmessage')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>r</div> ".$this->gettext('replytomessage')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>R</div> ".$this->gettext('replytoallmessage')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>s</div> ".$this->gettext('quicksearch')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>u</div> ".$this->gettext('checkmail')."<br class='clear' />";
        $c .= "<div class='shortcut_key'> </div> <br class='clear' />";
        $c .= "</div>";

        if(!is_object($rcmail->imap)){
          $rcmail->storage_connect();
        }
        $threading_supported = $rcmail->config->get('message-threading');
        $threading_supported = !empty($threading_supported);

        if ($threading_supported) {
          $c .= "<div><h4>".$this->gettext("threads")."</h4>";
          $c .= "<div class='shortcut_key'>E</div> ".$this->gettext('expand-all')."<br class='clear' />";
          $c .= "<div class='shortcut_key'>C</div> ".$this->gettext('collapse-all')."<br class='clear' />";
          $c .= "<div class='shortcut_key'>U</div> ".$this->gettext('expand-unread')."<br class='clear' />";
          $c .= "<div class='shortcut_key'> </div> <br class='clear' />";
          $c .= "</div>";
        }
        $c .= "<div><h4>".$this->gettext("messagesdisplaying")."</h4>";
        $c .= "<div class='shortcut_key'>d</div> ".$this->gettext('deletemessage')."<br class='clear' />";
        if (!empty($archive_supported)) {
          $c .= "<div class='shortcut_key'>z</div> ".$this->gettext('archive.buttontitle')."<br class='clear' />";
        }
        $c .= "<div class='shortcut_key'>c</div> ".$this->gettext('compose')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>f</div> ".$this->gettext('forwardmessage')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>i</div> ".$this->gettext('backtolist')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>j</div> ".$this->gettext('previousmessage')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>k</div> ".$this->gettext('nextmessage')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>p</div> ".$this->gettext('printmessage')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>r</div> ".$this->gettext('replytomessage')."<br class='clear' />";
        $c .= "<div class='shortcut_key'>R</div> ".$this->gettext('replytoallmessage')."<br class='clear' />";
        $c .= "<div class='shortcut_key'> </div> <br class='clear' />";
        $c .= "</div></div>";

        $rcmail->output->set_env('ks_functions', array('63' => 'ks_help'));

        $p['content'] = $c . $p['content'];
      }
      return $p;
    }
}


shortcuts.js
/**
 * Copyright 2014 Cor Bosman (cor@roundcu.be)
 *
 * roundcube shortcuts
 *
 * @author Cor Bosman <cor@roundcu.be>
 * @licence GNU GPL
 * @version 1.0
 */


(function(window, document, undefined)
{
    /**
     * registered shortcuts
     * @type {Object}
     */
    var _config = {};

    /**
     * all available commands
     * @type {Object}
     */
    var _all_commands = {};

    /**
     * current action
     * @type String
     */
    var _action = _get_action();

    /**
     * in settings keep track of the original key once you click on an input
     */
    var _oldkey;

    /**
     * all modifier keys
     * @type {Object}
     */
    var _MODIFIERS = {
        16: 'shift',
        17: 'ctrl',
        18: 'alt',
    };

    /**
     * forbidden keys, mostly because used in RC itself
     * @type {Object}
     */
    var _FORBIDDEN = {
        8: 'backspace',
        9: 'tab',
        10: 'enter',
        12: 'numlock',
        13: 'enter',
        20: 'capslock',
        27: 'esc',
        32: 'space',
        37: 'left',
        38: 'up',
        39: 'right',
        40: 'down',
        46: 'del',
    };

    /**
     * we resort to using an allowed keys object because cross-browser keycode detection is a huge mess
     * @type {Object}
     */
    var _ALLOWED = {
        33: 'PageUp',
        34: 'PageDown',
        35: 'End',
        36: 'Home',
        45: 'Insert',
        46: 'Delete',
        48: '0',
        49: '1',
        50: '2',
        51: '3',
        52: '4',
        53: '5',
        54: '6',
        55: '7',
        56: '8',
        57: '9',
        65: 'a',
        66: 'b',
        67: 'c',
        68: 'd',
        69: 'e',
        70: 'f',
        71: 'g',
        72: 'h',
        73: 'i',
        74: 'j',
        75: 'k',
        76: 'l',
        77: 'm',
        78: 'n',
        79: 'o',
        80: 'p',
        81: 'q',
        82: 'r',
        83: 's',
        84: 't',
        85: 'u',
        86: 'v',
        87: 'w',
        88: 'x',
        89: 'y',
        90: 'z',
        96: 'Num0',
        97: 'Num1',
        98: 'Num2',
        99: 'Num3',
        100: 'Num4',
        101: 'Num5',
        102: 'Num6',
        103: 'Num7',
        104: 'Num8',
        105: 'Num9',
        112: 'F1',
        113: 'F2',
        114: 'F3',
        115: 'F4',
        116: 'F5',
        117: 'F6',
        118: 'F7',
        119: 'F8',
        120: 'F9',
        121: 'F10',
        122: 'F11',
        123: 'F12',

    };

    /**
     * initialise
     */
    function _init() {
       /**
        * keypress listener
        */
       $(document).keydown(function (e) {
           return _handle_keypress(e);
       });
    }

    /**
     * handle keypress
     * @param  {Object} e
     * @return Boolean
     */
    function _handle_keypress(e)
    {

        // special case. If we hit ctrl-enter, and we're composing, and we have focus, then send email
        if (_action == 'compose' && (e.which == 13 || e.which == 10) && e.ctrlKey && $("*:focus").is("#composebody")) {
            $('.button.send').click();
            return false;
        }

        // check if some element has focus. If it does, skip this plugin.
        if ( $("*:focus").is("textarea, input") ) return true;

        // pressed keycode
        var keycode = _get_keycode(e);

        // skip if we're only hitting a modifier
        if(_is_modifier(keycode)) return true;

        // get the modifier
        var modifiers = _get_modifiers(e);

        // translate key to something readable
        var sequence = _translate(keycode, modifiers, e);

        // check the command arrays for a match
        if(typeof _config === 'object' && _config.shortcuts) {
            if(_action in _config.shortcuts && sequence in _config.shortcuts[_action]) {
                _run(_config.shortcuts[_action][sequence],_action);
                e.preventDefault();
                return false;
            } else if(sequence in _config.shortcuts.all) {
                _run(_config.shortcuts.all[sequence], 'all');
                e.preventDefault();
                return false;
            }
        }

        // we did not find a key, just pass the event along
        return true;
    }

    /**
     * find the keycode
     * @param  {Objecy} e
     * @return {int}
     */
    function _get_keycode(e)
    {
        e = e || window.event;
        return e && e.keyCode ? e.keyCode : (e && e.which ? e.which : 0);
    }

    /**
     * fire off the command
     * @param  {string} command
     */
    function _run(command, action)
    {
        var parameters;

        if(typeof Commands[command] === 'function') {
            Commands[command]();
        } else if(_config.commands[action].commands[command]) {

            // find parameters
            if(_config.commands[action].commands[command].parameters) {
                if(typeof _config.commands[action].commands[command].parameters == 'object') {
                    parameters = _config.commands[action].commands[command].parameters.join();
                } else {
                    parameters = _config.commands[action].commands[command].parameters;
                }
            }

            // do we have a different function to call?
            if(_config.commands[action].commands[command].function) {
                command = _config.commands[action].commands[command].function;
            }

            // execute
            rcmail.command(command, parameters);

        } else if(rcmail.commands[command]) {
            rcmail.command(command);
        } else {
            // we cant find a function that matches this
        }
    }

    /**
     * find the current action
     * @return {String}
     */
    function _get_action()
    {
        return rcmail.env.action === '' || rcmail.env.action == 'preview' || rcmail.env.action == 'show' ? 'mail' : rcmail.env.action;
    }

    /**
     * record a key press in settings
     * @param  {object} input jquery input object
     * @param  {object} k     keypress event
     * @return {boolean}
     */
    function _record_key(input,k) {

        // if we're recording, cancel all previous recordings
        _reset_recording();

        // save old key
        _oldkey = input.val();

        // set new recording
        input.attr('recording', 'true');

        // empty input
        input.val('');

        // ask for new key
        input.keydown(function(e) {

            // pressed keycode
            var keycode = _get_keycode(e);

            // skip if we're only hitting a modifier
            if(_is_modifier(keycode)) return false;

            // get the modifier
            var modifiers = _get_modifiers(e);

            // check if this is a valid key
            if((message = _validate_key(keycode, input, modifiers)) !== true) {
               return _record_fail(message,e);
            }

            // translate key to something readable
            var sequence = _translate(keycode, modifiers, e);

            // check if this sequence exists
            if(_key_exists(sequence, input)) {
                return _record_fail('keyexists',e);
            }

            // set the input field to this key sequence
            input.val(sequence);
            input.attr('value', sequence);

            // visually show we need to save
            $('.footerleft .button').css('color', 'orange');

            // unbind the keydown event
            $('.key').unbind('keydown')
                    .unbind('blur')
                    .removeAttr("recording")
                    .blur();


            // dont do anything else with this keypress
            e.preventDefault();

            return false;

        });

        // if we click someplace else, reset the recording.
        input.blur(function(i) {
            _reset_recording();
        });
    }

    /**
     * record failed
     * @param  {object} message
     * @param  {object} e
     * @return false
     */
    function _record_fail(message,e) {
        rcmail.display_message(rcmail.gettext(message, 'keyboard_shortcuts'), 'warning', 1500);
        _reset_recording();
        e.preventDefault();
        return false;
    }

    /**
     * capitalize a word
     * @param  {string} str
     * @return {string}
     */
    function capitalize(str) {
        return str.substr(0,1).toUpperCase() + str.substr(1).toLowerCase();
    }

    /**
     * check to see if a key is a modifier
     * @param  {string}  keycode
     * @return {Boolean}
     */
    function _is_modifier(keycode) {
        return _MODIFIERS[keycode] ? true : false;
    }

    /**
     * get all modifiers
     * @param  {object} e
     * @return {array}
     */
    function _get_modifiers(e) {
        var modifiers = [];

        if (e.shiftKey) {
            modifiers.push('shift');
        }

        if (e.altKey) {
            modifiers.push('alt');
        }

        if (e.ctrlKey) {
            modifiers.push('ctrl');
        }

        if (e.metaKey) {
            modifiers.push('meta');
        }

        return modifiers;
    }

    /**
     * translate a keycode to a readable string
     * @param  {string} keycode
     * @param  {array} modifiers
     * @param  {object} e
     * @return {string}
     */
    function _translate(keycode, modifiers, e) {

        // get character from keycode
        // var char = String.fromCharCode(keycode);
        var char = _ALLOWED[keycode];

        // if we pressed a letter, then return lowercase
        //if(keycode >= 65 || keycode <= 90) {
        //    char = char.toLowerCase();
        //}

        // do we have a modifier?
        var modifier = modifiers.length > 0 ? capitalize(modifiers[0])+'+' : '';

        return modifier+char;
    }

    /**
     * validate a recorded key
     * @param  {object} i
     * @return true|String
     */
    function _validate_key(keycode, input, modifiers) {

        if(keycode === false) {
            return 'invalidkey';
        }

        // skip keys that roundcube itself uses
        if(_FORBIDDEN[keycode]) {
            return 'invalidkey';
        }

        if(!_ALLOWED[keycode]) {
            return 'invalidkey';
        }

        // we dont allow multiple modifiers
        if(modifiers.length > 1) {
            return 'multiple_modifiers';
        }

        return true;
    }

    /**
     * check if the key exists
     * @param  {string} keycode
     * @param  {object} input Jquery object
     * @return {boolean}
     */
    function _key_exists(sequence, input) {
        var found = false;

        // current section
        var section = input.attr('data-section');

        // which row did we record on
        var row = input.closest('tr');

        // see if we have the same key in this section or in all
        input.closest('form').find("input.key[value='"+sequence+"']").each(function(e) {
            // section that has a match
            var cur_section = $(this).attr('data-section');

            // row that has a match
            var cur_row = $(this).closest('tr');

            // is this key allowed?
            if(!row.is(cur_row)) {
                if(section == cur_section || cur_section == 'all' || section == 'all') {
                    found = true;
                }
            }
        });
        return found;
    }

    /**
     * reset the recording
     * @return {}
     */
    function _reset_recording() {
        $(".key[recording='true']").val(_oldkey)
                                       .removeAttr('recording').blur();
    }

    /**
     * functions for commands where we cant use internal roundcube commands
     */
    var Commands = {

        /**
         * show help
         */
        help: function() {
            if($('.keyboard_shortcuts_help').length) {
                $('.keyboard_shortcuts_help').dialog('open');
            } else {
                rcmail.http_request('plugin.ks_get_help', false);
            }
        },

        // search
        search: function() {
            $('#quicksearchbox').focus();
            $('#quicksearchbox').select();
        },

    };

    /**
     * initialize
     */


    /**
     * shortcuts object for outside access
     * @type {Object}
     */
    var Shortcuts = {
        /**
         * set the user shortcuts
         * @param  {Object} shortcuts
         */
        config: function(config) {
            $.extend(_config, config);
        },

        /**
         * allow access to help function for icon click event
         */
        help: function() {
            Commands.help();
        },

        /**
         * initialise the settings screen
         */
        init: function() {
            _init();
        },

        /**
         * record key in settings
         * @param  {object} input
         * @param  {object} e
         */
        record: function(input,e) {
            _record_key(input,e);
        }
    };

    // expose shortcuts object to the outside
    window.Shortcuts = Shortcuts;


}) (window, document);


